manual/manual/tutorials/coreexamples.etex

```diff
@@ -677,6 +677,89 @@ print_expr e; print_newline ();;
 print_expr (deriv e "x"); print_newline ();;
 \end{caml_example}
 
+\section{Printf formats}
+
+There is a "printf" function in
+\href{https://caml.inria.fr/pub/docs/manual-ocaml/libref/Printf.html}{Printf}
+module (see chapter~\ref{c:moduleexamples}) that allows you to make formatted
+output more concisely.
+It follows the behavior of the "printf" function from the C standard library.
+The "printf" function takes a format string that describes the output values and
+their format, and arguments for output that correspond to the formatting
+directives in the format string.
+\begin{caml_example}{toplevel}
+Printf.printf "%i + %i is an integer value, %F * %F is a float, %S\n"
+3 2 4.5 1. "this is a string";;
+\end{caml_example}
+The OCaml type system checks that the type of the arguments and the specifiers are
+compatible. If you pass it an argument of a type that does not correspond to
+the format specifier, the compiler will display an error message:
+\begin{caml_example}{toplevel}[error]
+Printf.printf "Float value: %F" 42;;
+\end{caml_example}
+Associated with this feature is the need to explicitly indicate that the
+string is a format string in order to pass it to the "printf" function.
+\begin{caml_example*}{toplevel}
+let str : _ format =
+    "%i is an integer value, %F is a float, %S\n";;
+\end{caml_example*}
+\begin{caml_example}{toplevel}
+Printf.printf str 3 4.5 "string value";;
+\end{caml_example}
```

Comment:
Personally, I would move this example at the end of the section after the `print_expression` example since knowing how to store format string in variables feels less crucial than knowing how to define composable printers.

Reply:
In my opinion, it is worth going from the simpler to the complex, which is why I left this example before starting the description of composable printers.

Reply:
My opinion is precisely that this example is much more complex and uncommon than composable printers. In particular, the definition of `str` yields the following toplevel output:
```OCaml
# let str: _ format = "%i is an integer value, %F is a float, %S\n";;

val str : (int -> float -> string -> 'a, 'b, 'a) format =
  CamlinternalFormatBasics.Format
   (CamlinternalFormatBasics.Int (CamlinternalFormatBasics.Int_i,
     CamlinternalFormatBasics.No_padding,
     CamlinternalFormatBasics.No_precision,
     CamlinternalFormatBasics.String_literal (" is an integer value, ",
      CamlinternalFormatBasics.Float (CamlinternalFormatBasics.Float_F,
       CamlinternalFormatBasics.No_padding,
       CamlinternalFormatBasics.No_precision,
       CamlinternalFormatBasics.String_literal (" is a float, ",
        CamlinternalFormatBasics.Caml_string
         (CamlinternalFormatBasics.No_padding,
         CamlinternalFormatBasics.Char_literal ('\n',
          CamlinternalFormatBasics.End_of_format)))))),
   "%i is an integer value, %F is a float, %S\n")
```
which is quite esoteric-looking for  … anybody not familiar with the implementation of format strings. You rightfully elided this output, but the complexity is still here.

Reply:
Indeed, a fair point. It is probably worthwhile to move this example to the end of the section, but leave the output hidden. For the first acquaintance with the language, it is probably enough to know about the existence of the ability to store strings in variables, but not necessarily to get acquainted with the implementation details.